flow/activities/maintenance_activity.go

```diff
@@ -0,0 +1,290 @@
+			fmt.Errorf("error decoding mirror status for maintenance: %w", err)
+	}
+	return state, nil
+}
+
+func (a *MaintenanceActivity) WaitForRunningSnapshots(ctx context.Context) (MaintenanceMirrorsInfo, error) {
+	mirrors, err := a.GetAllMirrors(ctx)
+	if err != nil {
+		return MaintenanceMirrorsInfo{}, err
+	}
+
+	slog.Info("Found mirrors for snapshot check", "mirrors", mirrors, "len", len(mirrors.Mirrors))
+
+	waitBeforeAlertingDuration := 30 * time.Minute
+
+	for _, mirror := range mirrors.Mirrors {
+		lastStatus, err := a.checkAndWaitIfSnapshot(ctx, mirror, 2*time.Minute, waitBeforeAlertingDuration)
+		if err != nil {
+			return MaintenanceMirrorsInfo{}, err
+		}
+		slog.Info("Finished checking and waiting for snapshot",
+			"mirror", mirror.Name, "workflowId", mirror.WorkflowId, "lastStatus", lastStatus.String())
+	}
+	slog.Info("Finished checking and waiting for all mirrors to finish snapshot")
+	return mirrors, nil
+}
+
+func (a *MaintenanceActivity) checkAndWaitIfSnapshot(
+	ctx context.Context,
+	mirrorInfo MaintenanceMirrorInfoItem,
+	logEvery time.Duration,
+	alertEvery time.Duration,
+) (protos.FlowStatus, error) {
+	// In case a mirror was just kicked off, it shows up in the running state, we wait for a bit before checking for snapshot
+	if mirrorInfo.CreatedAt.After(time.Now().Add(-30 * time.Second)) {
+		slog.Info("Mirror was created less than 30 seconds ago, waiting for it to be ready before checking for snapshot",
+			"mirror", mirrorInfo.Name, "workflowId", mirrorInfo.WorkflowId)
+		time.Sleep(30 * time.Second)
+	}
+
+	mirrorStatus, err := a.getMirrorStatus(ctx, mirrorInfo)
+	if err != nil {
+		return mirrorStatus, err
+	}
+
+	slog.Info("Checking if mirror is snapshot", "mirror", mirrorInfo.Name, "workflowId", mirrorInfo.WorkflowId, "status", mirrorStatus.String())
+	if mirrorStatus != protos.FlowStatus_STATUS_SNAPSHOT && mirrorStatus != protos.FlowStatus_STATUS_SETUP {
+		return mirrorStatus, nil
+	}
+	slog.Info("Waiting for mirror to finish snapshot",
+		"mirror", mirrorInfo.Name, "workflowId", mirrorInfo.WorkflowId, "status", mirrorStatus.String())
+	defer shared.Interval(ctx, alertEvery, func() {
+		slog.Warn("[Maintenance] Still waiting for mirror to finish snapshot",
+			"mirror", mirrorInfo.Name, "workflowId", mirrorInfo.WorkflowId, "status", mirrorStatus.String())
+		a.Alerter.LogNonFlowWarning(ctx, telemetry.MaintenanceWait, mirrorInfo.Name, fmt.Sprintf(
+			"Maintenance mode is still waiting for mirror to finish snapshot, mirror=%s, workflowId=%s, status=%s",
+			mirrorInfo.Name, mirrorInfo.WorkflowId, mirrorStatus))
+	})()
+
+	defer shared.Interval(ctx, logEvery, func() {
+		slog.Info("[Maintenance] Waiting for mirror to finish snapshot",
+			"mirror", mirrorInfo.Name, "workflowId", mirrorInfo.WorkflowId, "status", mirrorStatus.String())
+	})()
+
+	snapshotWaitSleepInterval := 10 * time.Second
+	for mirrorStatus == protos.FlowStatus_STATUS_SNAPSHOT || mirrorStatus == protos.FlowStatus_STATUS_SETUP {
```

Comment:
waiting for a certain `mirrorStatus` can be made a helper

Reply:
It can happen that a mirror got paused by the time we were waiting. We just need to make sure that things are not in these 2 states, rest we will take care of pausing later

Reply:
I think it should be fine for now, if it was at a third place we can extract